rln/src/merkle_tree.rs

```diff
@@ -1,132 +1,382 @@
-// Implementation adapted from https://github.com/worldcoin/semaphore-rs/blob/d462a4372f1fd9c27610f2acfe4841fab1d396aa/src/merkle_tree.rs
-// In our customization, we expand MerkleTree to have a next_index counter, so that we can add a public API to add leaves to the next available counter with no need to specify the index
+// This crate provides different implementation of Merkle tree
+// Currently two interchangeable implementations are supported:
+//    - FullMerkleTree: each tree node is stored
+//    - OptimalMerkleTree: only nodes used to prove accumulation of set leaves are stored
```

Comment:
I'm confused by "optimal" name, where is this from? is this not just IncrementalMerkleTree? Or perhaps Sparse MT

Reply:
ah my bad, I confused incremental and optimal based on how I thought kilic used incremental mt algo